neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,462 @@
+    /// Returns true if the packet is a PMTUD probe.
+    #[must_use]
+    pub fn is_pmtud_probe(&self, p: &SentPacket) -> bool {
+        self.probe_state == Probe::Sent && p.len() == self.probe_size()
+    }
+
+    /// Count the PMTUD probes included in `pkts`.
+    fn count_pmtud_probes(&self, pkts: &[SentPacket]) -> usize {
+        pkts.iter().filter(|p| self.is_pmtud_probe(p)).count()
+    }
+
+    /// Checks whether a PMTUD probe has been acknowledged, and if so, updates the PMTUD state.
+    /// May also initiate a new probe process for a larger MTU.
+    pub fn on_packets_acked(&mut self, acked_pkts: &[SentPacket], stats: &mut Stats) {
+        let acked = self.count_pmtud_probes(acked_pkts);
+        if acked == 0 {
+            return;
+        }
+        // A probe was ACKed, confir, the new MTU and try to probe upwards further.
+        stats.pmtud_ack += acked;
+        self.mtu = self.search_table[self.probe_index];
+        qdebug!("PMTUD probe of size {} succeeded", self.mtu);
+        self.start_pmtud();
+    }
+
+    /// Stops the PMTUD process, setting the MTU to the largest successful probe size.
+    fn stop_pmtud(&mut self, idx: usize, now: Instant) {
+        self.probe_state = Probe::NotNeeded; // We don't need to send any more probes
+        self.probe_index = idx; // Index of the last successful probe
+        self.mtu = self.search_table[idx]; // Leading to this MTU
+        self.probe_count = 0; // Reset the count
+        self.loss_counts = vec![0; self.search_table.len()]; // Reset the loss counts
+        self.raise_timer = Some(now + PMTU_RAISE_TIMER);
+        qdebug!(
+            "PMTUD stopped, PLPMTU is now {}, raise timer {:?}",
+            self.mtu,
+            self.raise_timer.unwrap()
+        );
+    }
+
+    /// Checks whether a PMTUD probe has been lost. If it has been lost more than `MAX_PROBES`
+    /// times, the PMTUD process is stopped.
+    pub fn on_packets_lost(
+        &mut self,
+        lost_packets: &[SentPacket],
+        stats: &mut Stats,
+        now: Instant,
+    ) {
+        // Track lost probes
+        let lost = self.count_pmtud_probes(lost_packets);
+        stats.pmtud_lost += lost;
+
+        // Increase loss counts for all sizes included in the lost packets.
+        for (count, inc) in self.loss_counts.iter_mut().zip(
+            self.search_table
+                .iter()
+                .map(|len| lost_packets.iter().filter(|p| p.len() > *len).count()),
+        ) {
+            *count += inc;
+        }
+
+        // Check if any packet of size > MTU has been lost MAX_PROBES times or more.
+        let Some(last_good) = self.loss_counts.iter().rposition(|&c| c >= MAX_PROBES) else {
+            // If not, keep going.
+            if lost > 0 {
+                // Don't stop the PMTUD process.
+                self.probe_state = Probe::Needed;
+            }
+            return;
+        };
+
+        qdebug!("Packet of size > {} lost {} times", self.mtu, MAX_PROBES);
```

Comment:
```suggestion
        qdebug!("Packet of size > {} lost >= {} times", self.mtu, MAX_PROBES);
```
 Would be more correct, right?